### PR TITLE
Draft PR Adding mistral 0.1

### PR DESCRIPTION
**Pull request by AIproj**
Here's the PR of the October addition of support for Mistral 7B v0.1 in GPT-NeoX, referred to in [issue 1050](https://github.com/EleutherAI/gpt-neox/issues/1050). 

Among other things, this PR also adds support for sliding window attention in GPT-NeoX, both through FlashAttention2 and through Megatron.

An example script is included to show how to run the conversion of a HuggingFace (HF) Mistral 7B v0.1 model into corresponding GPT-NeoX checkpoints.

The items left to do since then are to:
- Add support for PipelineEngine in the HF -> GPT-NeoX conversion script (currently, `pp>0` is not supported).
- Test training through HF to check that on enwik8 the loss is also ~3 going down quickly to 2.xx.
- Try to lm-eval the GPT-NeoX and HF versions to see if they match in performance.

Note: [issue 1124](https://github.com/EleutherAI/gpt-neox/pull/1124) recently added support for conversion back to HF to enable such testing and is also concerned with supporting pp>0 in the conversion scripts.

### CODE REVIEW

**Comment by haileyschoelkopf** [comment]
Thank you @AIproj , I can finish out from here!

File: configs/mistral_7b.yml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  "checkpoint_factor": 250,
+  "eval_interval": 250,
+  "eval_iters": 25,
+
+  "log_interval": 1,
+  "steps_per_print": 1,
+  "wall_clock_breakdown": true,
+
+  "tokenizer_type": "SPMTokenizer",
+  "vocab-file": "/path/to/mistral-7B-v0.1/tokenizer.model", # use tokenizer.model from Meta CodeLlama download
```
Comment (StellaAthena) [inline configs/mistral_7b.yml:94]: This appears to be a held-over comment from when this file was used for CodeLLaMA

Comment (StellaAthena) [inline configs/mistral_7b.yml:94]: The comment here appears to be a held-over comment from when this file was used for LLeMMA

File: configs/mistral_7b.yml [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  # "load": "" # set to same as "save" to resume from intermediate finetuning step
+  "load": "checkpoints/neox-converted",
+
+# Set these 2 (finetune and validation with fw pass) to false during conversion
+  "finetune": false, # set to false once resuming from intermediate finetuning step
+  "checkpoint_validation_with_forward_pass": false,
+
+
+  "use_wandb": false,
+  "wandb_group": "llemma_34b_replication",
```
Comment (StellaAthena) [inline configs/mistral_7b.yml:106]: The naming here appears to be a held-over comment from when this file was used for LLeMMA

File: megatron/neox_arguments/neox_args.py [hunk 3]
```diff
@@ -105,6 +105,10 @@ class NeoXArgsModel(NeoXArgsTemplate):
     Transformer hidden size.
     """
 
+    intermediate_size: int = None
```
Comment (StellaAthena) [inline megatron/neox_arguments/neox_args.py:108]: We should include in the help string here that this is (in general) discouraged and that the code will pick something nice for you if you omit it.

File: configs/mistral_7b.yml [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  "log_interval": 1,
+  "steps_per_print": 1,
+  "wall_clock_breakdown": true,
+
+  "tokenizer_type": "SPMTokenizer",
+  "vocab-file": "/path/to/mistral-7B-v0.1/tokenizer.model", # use tokenizer.model from Meta CodeLlama download
+
+  "save": "checkpoints/test",
+  # "load": "" # set to same as "save" to resume from intermediate finetuning step
+  "load": "checkpoints/neox-converted",
```
Comment (StellaAthena) [inline configs/mistral_7b.yml:98]: Let's change this to match `save` before someone gets mess up by it.

File: README.md [hunk 5]
```diff
@@ -1,3 +1,8 @@
+# Work in Progress: adding support for Mistral
```
Comment (StellaAthena) [inline README.md:1]: We should remove this text now that we're moving it up to an almost main branch xD

Comment (haileyschoelkopf) [reply README.md:1]: To clarify, yes this PR will be updated in order to merge it to the main branch along with Llama 2 (GQA) support, docs + clean configs and all!

**Comment by haileyschoelkopf** [comment]
This is ready for review! 

Something might be up with the self-hosted runner for tests? it seems to not have the proper packages installed, including pytest

**Review by Quentin-Anthony — APPROVED** [review]

**Comment by malteos** [comment]
Does this require flash attention >= 2.3? Sliding window attention is only available from that version (see https://github.com/Dao-AILab/flash-attention?tab=readme-ov-file#23-local-ie-sliding-window-attention).

With the current Docker image (`flash-attn==2.2.1`) I get the following error:

```
...
  File "/netscratch/mostendorff/experiments/gpt-neox/megatron/model/transformer.py", line 622, in flash_attention
    output = self.flash_qkv_fn(
TypeError: flash_attn_func() got an unexpected keyword argument 'window_size'
```

**Comment by haileyschoelkopf** [comment]
Ah yes you’re correct— I will add a check for this.

**Pull request merged**